snuba/migrations/operations.py

```diff
@@ -0,0 +1,41 @@
+from snuba.clusters.cluster import ClickhouseClientSettings, get_cluster
+from snuba.clusters.storage_sets import StorageSetKey
+
+
+class Operation:
```

Comment:
What is the conceptual difference between a Migration and an Operation ? In other words, when would I create multiple migrations and when instead would I create multiple operations within one migration ?
Asking because unless there is a conceptual distinction we can make apparent (like "Operations are individual statements that the user is not supposed to extend" this is just an example) we should probably not have the two. But, again, not sure what your plan is for this.

Reply:
I think migrations are a set of operations (often one but can be more) that can be logically grouped or where the intermediate state between operations may not be considered to be valid. Putting them in a same migration will ensure they are always executed together.

I think often the decision to split operations over separate migrations or group them together won't really matter and can be up to the person writing the migration. However some situations where I think we can group operations together are:
- providing the initial schema migration to recreate the current state
- creating a new table and filling it with data
- creating a table and a materialized view that uses that table
- adding a field like "message_timestamp" or "flattened_tags" to a number of different tables at once

Reply:
I think the consistency argument is by far the strongest and should be documented in the docstring of the class:
"where the intermediate state between operations may not be considered to be valid"

Essentially does this mean that a migration brings the system from a consistent state to another in which the system can function while an operation is an intermediate step that can bring the system in a reversible potentially inconsistent state ? If yes how do we enforce we can always go back to a consistent state ?

Anyway the distinction should be clear and documented. Having not well defined components with an ambiguous responsibility often opens the door to misuse.